database-switching-challenge/main/java/org/jakartaee/sample/websocket/GameSessions.java

```diff
@@ -0,0 +1,265 @@
+
+@ApplicationScoped
+public class GameSessions {
+
+    private static final Logger LOG = Logger.getLogger(GameSessions.class.getName());
+
+    private final Map<String, Session> sessionsById = new HashMap<>();
+
+    private final Map<String, Player> playersBySessionId = new HashMap<>();
+
+    private final Game game = new Game();
+
+    private Event<GameEvent> gameEventPublisher;
+    private Event<GameState> gameStatePublisher;
+
+    @Inject
+    public GameSessions(Event<GameEvent> gameEventPublisher,
+                        Event<GameState> gameStatePublisher) {
+        this.gameEventPublisher = gameEventPublisher;
+        this.gameStatePublisher = gameStatePublisher;
+    }
+
+    /**
+     * do not use it, it is required by CDI
+     */
+    @Deprecated
+    public GameSessions() {
+    }
+
+    @PostConstruct
+    void postConstruct() {
+        game.addListener(this::onGameEvent)
+                .addListener(gameEventPublisher::fire)
+                .addListener(event -> gameStatePublisher.fire(event.gameState()));
+    }
+
+    private Optional<Player> getPlayerBySession(Session session) {
+        return getPlayerBySessionId(session.getId());
+    }
+
+    private Optional<Player> getPlayerBySessionId(String sessionId) {
+        synchronized (this) {
+            return Optional.ofNullable(playersBySessionId.get(sessionId));
+        }
+    }
+
+    private Optional<Session> getSessionById(String sessionId) {
+        synchronized (this) {
+            return Optional.ofNullable(sessionsById.get(sessionId));
+        }
+    }
+
+    public void register(Session session, String playerName) {
+        synchronized (this) {
+            String sessionId = session.getId();
+            sessionsById.put(sessionId, session);
+            playersBySessionId.put(sessionId, Player.of(sessionId, playerName));
+        }
+        send(session, CONNECTED.build());
+    }
+
+    public void unregister(Session session) {
+
+        synchronized (this) {
+            String sessionId = session.getId();
+            sessionsById.remove(sessionId, session);
+            leaveGame(Optional.ofNullable(playersBySessionId.remove(sessionId)), session);
+        }
+    }
+
+    private void leaveGame(Optional<Player> playerRef, Session session) {
+        playerRef
+                .ifPresent(game::leaveGame);
+    }
+
+    public void playerWantsNewGame(Message message, Session session) {
+        getPlayerBySession(session)
+                .ifPresent(game::newGame);
+    }
+
+    public void playerWantsToPlay(Message message, Session session) {
+        getPlayerBySession(session)
+                .ifPresent(player -> movement.get(message)
+                        .map(Movement::valueOf)
+                        .ifPresent(playerMovement -> game.playGame(player, playerMovement)));
+    }
+
+    private void onGameEvent(GameEvent event) {
+
+        var gameState = event.gameState();
+
+        if (gameState instanceof GameInvalid gameInvalid
```

Comment:
Have you tried the State pattern here?

Reply:
Yeah, it's a kind of! I implemented the game where each game match is a machine-state, then when the game state changes, a listener is called